metadetect/lsst_measure_scarlet.py

```diff
@@ -0,0 +1,992 @@
+    psf_jacob = jacob.copy()
+    psf_jacob.set_cen(row=psf_cen[0], col=psf_cen[1])
+
+    # we will have need of the bit names which we can only
+    # get from the mask object
+    # this is sort of monkey patching, but I'm not sure of
+    # a better solution
+    meta = {'maskobj': maskobj}
+
+    psf_obs = ngmix.Observation(
+        psf_im,
+        weight=psf_wt,
+        jacobian=psf_jacob,
+    )
+    obs = ngmix.Observation(
+        im,
+        weight=wt,
+        bmask=bmask,
+        jacobian=jacob,
+        psf=psf_obs,
+        meta=meta,
+    )
+
+    return obs
+
+
+def _extract_jacobian(wcs, subim, source, orig_cen):
+    """
+    extract an ngmix.Jacobian from the image object
+    and object record
+
+    Parameters
+    ----------
+    wcs: stack wcs
+        The wcs for the full image not this sub image
+    imobj: lsst.afw.image.Exposure
+        An Exposure object, e.g. ExposureF
+    source: lsst.afw.table.SourceRecord
+        A source record from detection/deblending
+    orig_cen: Point2D
+        Location of object in original image
+
+    returns
+    --------
+    Jacobian: ngmix.Jacobian
+        The local jacobian
+    """
+
+    xy0 = subim.getXY0()
+    stamp_cen = orig_cen - geom.Extent2D(xy0)
+
+    LOG.debug('cen in subim: %s', stamp_cen)
+    row = stamp_cen.getY()
+    col = stamp_cen.getX()
+
+    # we get this at the original center
+    linear_wcs = wcs.linearizePixelToSky(
+        orig_cen,  # loc in original image
+        geom.arcseconds,
+    )
+    jmatrix = linear_wcs.getLinear().getMatrix()
+
+    jacob = ngmix.Jacobian(
+        row=row,
+        col=col,
+        dudcol=jmatrix[0, 0],
+        dudrow=jmatrix[0, 1],
+        dvdcol=jmatrix[1, 0],
+        dvdrow=jmatrix[1, 1],
+    )
+
+    return jacob
+
+
+def _extract_weight(subim):
+    """
+    TODO get the estimated sky variance rather than this hack
+    TODO should we zero out other bits?
+
+    extract a weight map
+
+    Areas with NO_DATA will get zero weight.
+
+    Because the variance map includes the full poisson variance, which
+    depends on the signal, we instead extract the median of the parts of
+    the image without NO_DATA set
+
+    parameters
+    ----------
+    subim: sub exposure object
+    """
+
+    # TODO implement bit checking
+    var_image = subim.variance.array
+
+    weight = var_image.copy()
+
+    weight[:, :] = 0
+
+    wuse = np.where(var_image > 0)
+
+    if wuse[0].size > 0:
+        medvar = np.median(var_image[wuse])
+        weight[:, :] = 1.0/medvar
+    else:
+        print('    weight is all zero, found '
+              'none that passed cuts')
+
+    return weight
+
+
+def _extract_psf_image(exposure, orig_cen):
+    """
+    get the psf associated with this image
+
+    coadded psfs are generally not square, so we will
+    trim it to be square and preserve the center to
+    be at the new canonical center
+
+    TODO: should we really trim the psf to be even?  will this
```

Comment:
I'd argue we always want the PSF in an odd image and drawn at a pixel center. Not sure what is going on here.

Reply:
Yeah, this is to deal with the fact that DM PSF image shapes vary wildlly and are non square.  Probably should be rewritten to pad to odd in some sensible way.

Reply:
Might be better to assert we have a square psf image on input, and push this upstream to the coadd psf code.   That code actually takes the psf dims from the user, so we are safe as long as we are running our code.  Will have to revisit if DM rewrites our coadd code.

Reply:
(assert square and odd)